invenio_rdm_records/services/pids/providers/datacite.py

```diff
@@ -178,10 +185,20 @@ def update(self, pid, record, url=None, **kwargs):
         :param record: the record metadata for the DOI.
         :returns: `True` if is updated successfully.
         """
+        hide = False
+        if isinstance(record, ChainObject):
+            if record._child["access"]["record"] == "restricted":
+                hide = True
+        elif record["access"]["record"] == "restricted":
+            hide = True
+
         try:
-            # Set metadata
-            doc = self.serializer.dump_obj(record)
-            self.client.api.update_doi(metadata=doc, doi=pid.pid_value, url=url)
+            if hide:
+                self.client.api.hide_doi(doi=pid.pid_value)
+            else:
+                doc = self.serializer.dump_obj(record)
```

Comment:
We reviewed this with @zzacharo and we think that when you change from restricted to public, here we are not calling `.register()`, so the local PID will still have the status `PIDStatus.NEW` or `PIDStatus.RESERVED`.
We are missing a test that will test the change from restricted to public and viceversa.

Reply:
the pid register method is already called, on the publish. that is independent of the way datacite is handling the restriction access setting.

Reply:
Yes, that call won't happen until the record is public. The "doi" pid doesn't need to be registered until the DOI is actually sent to DataCite. So every time there are changes to the record it will try to register the DOI, but it will only get registered if the record is public. A test is probably a good idea though.

Reply:
> What about this [call](https://github.com/inveniosoftware/invenio-rdm-records/pull/1724/files#diff-804dbcc4c6b21d366befbb2b80d5f987c616878f680562143f8999bef3e7a0e8R163). That will not be called if the record is restricted. Isn't this suppose to register the "doi" PID?

what do you mean by "register" the doi PID? 

as Tom said, the doi should only be set to R in `pidstore_pid` if it has been registered at datacite. and that isn't the case if the record is set to `restricted`. this means that the call in the [register_or_update](https://github.com/inveniosoftware/invenio-rdm-records/blob/f1d5dc0da0c71d29fd21696b7d25af79dd5d14b3/invenio_rdm_records/services/pids/service.py#L210) task will register the pid DOI at datacite after the change from `restricted` to `public`

> We reviewed this with @zzacharo and we think that when you change from restricted to public, here we are not calling `.register()`, so the local PID will still have the status `PIDStatus.NEW` or `PIDStatus.RESERVED`. We are missing a test that will test the change from restricted to public and viceversa.

is [this](https://github.com/inveniosoftware/invenio-rdm-records/pull/1724/files#diff-4d69e1b21e25061eb03ca154d0f090414925e0529fede5b88a2b6044d8d9e913R88) test what you had in mind?

Reply:
Maybe add  `minimal_record["access"]["record"] = "public"` at the end and then `assert pid.status == PIDStatus.REGISTERED` ?

Reply:
like that: https://github.com/inveniosoftware/invenio-rdm-records/pull/1724/commits/fe042f84c78ecda19197a94dab05f17b7e99d981

Reply:
Looks good to me

Reply:
Just silly copyright conflict to fix https://github.com/inveniosoftware/invenio-rdm-records/pull/1724/conflicts

Reply:
so rebased and cleaned, lets see tomorrow for the rest :)

Reply:
@utnapischtim thanks for the test, it seems good! I understand that the `PIDComponent` will register the doi PID, but I think before the datacite client when the method `client.register()` was called, it ensured that the local doi PID i.e PersitentIdentifier model was registered before publishing it to datacite. Now, that a "register" action can happen in the update and if the client is used directly e.g an admin is using the client programmatically then the doi will not be registered just by calling `client.update()` to a public record. I am thinking that we need to call 

```python
 local_success = super().register(pid)
 if not local_success:
     return False
```

in the case of the `event=publish`.

Reply:
i am not sure if i understand your scenario:

The `register` action in `update` can only happen if the `PersistentIdentifier` has been called in `dataciteprovicer.register` method. the register method will always be called first because of the check in the `register_or_update` method [see](https://github.com/inveniosoftware/invenio-rdm-records/blob/f1d5dc0da0c71d29fd21696b7d25af79dd5d14b3/invenio_rdm_records/services/pids/service.py#L210). since the DOI PID was not registered locally in the database if it was restricted the `pid.is_registered` will be false and therefore the `pid_manager.register` method will be called and therefore the `DatacitePIDProvider.register` method

do you mean that before of the `event:publish` the `DatacitePIDProvider.update` method would have failed because it the doi has not been registered at Datacite?

Reply:
I mean using the `DatacitProvider` directly without going through the service :) It is not the usual/recommended way but as an admin you might do that and I would rather have an extra check to register the local PID doi before going to datacite. The normal flow via services is working correctly no problem with that!

Reply:
then yes you are right, we have to add a if to register the pid if it has not been registered at that point. which makes the whole code really messy

Reply:
the `pid.register()` method already checks if a pid is registered otherwise do nothing.... Maybe I am overthinking it but I was just triggered by the fact that such a check was present in the register method..

Reply:
to be honest, in my point of view it is wrong to register the DOI pid in `pidstore_pid` within the provider class. the provider should only take care what's associated with the provider. so the `DataCitePIDProvider.register` should not know about the internal behavior and should only be called if it has been registered internally. But that is another discussion.

yes `pid.register` takes care of knowing if it has already bin registered but this is not called in `DataCitePIDProvider.update` as you pointed out. So if a record is set to `RESTRICTED` the DOI PID will not be set to R in `pidstore_pid`. If someone is using the `DataCitePIDProvider.update` method then directly the DOI will be registered at DataCite and will remain `K` in `pidstore_pid`. if the record is then set to `public` the code will mostly work, because the `public_doi` method of the datacite rest client does more or less the same as the update does, it uses the `event: publish` to register the doi.

the thing is, is it a problem if the doi is in state `K` in `pidstore_pid` for this scenario?

Reply:
I think the /doi link won't work if the identifier is in state `K`. I guess we could add a check to pidstore and then register if needed, though it's a lot of extra calls to take care of someone who's not really following the normal flow.